src/games/strategy/triplea/image/UnitImageFactory.java

```diff
@@ -107,60 +111,77 @@ private void clearImageCache() {
     if (m_images.containsKey(fullName)) {
-      return m_images.get(fullName);
+      return Optional.of(m_images.get(fullName));
     }
-    final Image baseImage = getBaseImage(baseName, player);
-    // We want to scale units according to the given scale factor.
-    // We use smooth scaling since the images are cached to allow
-    // to take our time in doing the scaling.
-    // Image observer is null, since the image should have been
-    // guaranteed to be loaded.
-    final int width = (int) (baseImage.getWidth(null) * m_scaleFactor);
-    final int height = (int) (baseImage.getHeight(null) * m_scaleFactor);
-    final Image scaledImage = baseImage.getScaledInstance(width, height, Image.SCALE_SMOOTH);
-    // Ensure the scaling is completed.
-    try {
-      Util.ensureImageLoaded(scaledImage);
-    } catch (final InterruptedException ex) {
-      ex.printStackTrace();
+    try (PerfTimer timer = Perf.startTimer("Load an image")) {
+      final Optional<Image> image = getBaseImage(baseName, player);
+      if (!image.isPresent()) {
+        return Optional.empty();
+      }
+      Image baseImage = image.get();
+
+
+      // We want to scale units according to the given scale factor.
+      // We use smooth scaling since the images are cached to allow
+      // to take our time in doing the scaling.
+      // Image observer is null, since the image should have been
+      // guaranteed to be loaded.
+      final int width = (int) (baseImage.getWidth(null) * m_scaleFactor);
+      final int height = (int) (baseImage.getHeight(null) * m_scaleFactor);
+      final Image scaledImage = baseImage.getScaledInstance(width, height, Image.SCALE_SMOOTH);
+      // Ensure the scaling is completed.
+      try {
+        Util.ensureImageLoaded(scaledImage);
+      } catch (final InterruptedException ex) {
+        ex.printStackTrace();
+      }
+      m_images.put(fullName, scaledImage);
+      return Optional.of(scaledImage);
     }
-    m_images.put(fullName, scaledImage);
-    return scaledImage;
   }
 
-  public URL getBaseImageURL(final String baseImageName, final PlayerID id) {
+  public Optional<URL> getBaseImageURL(final String baseImageName, final PlayerID id) {
     return getBaseImageURL(baseImageName, id, m_resourceLoader);
   }
 
-  public static URL getBaseImageURL(final String baseImageName, final PlayerID id,
+  private static Optional<URL> getBaseImageURL(final String baseImageName, final PlayerID id,
       final ResourceLoader resourceLoader) {
     // URL uses '/' not '\'
     final String fileName = FILE_NAME_BASE + id.getName() + "/" + baseImageName + ".png";
     final URL url = resourceLoader.getResource(fileName);
     if (url == null) {
-      throw new IllegalStateException("Cant load: " + baseImageName + "  looking in: " + fileName);
+      ClientLogger.logError("MISSING IMAGE (this unit or image will be invisible): " + baseImageName + "  looking in: " + fileName);
```

Comment:
The logError call will throw a pop-up informing the user.  This is the right place to log it, when we introduce the Optional.empty() in the first place. There are multiple independent higher levels that call into this code, they all would have to do the same thing if we logged the problem there.